events/config.go

```diff
@@ -15,66 +15,56 @@
+	ErrProviderNotConfigured = errors.New("events provider not configured")
+)
 
-// SubscriberConfig handles reading in all the config values available for setting up a pubsub publisher
-type SubscriberConfig struct {
-	URL        string        `mapstructure:"url"`
-	Timeout    time.Duration `mapstructure:"timeout"`
-	Prefix     string        `mapstructure:"prefix"`
-	QueueGroup string        `mapstructure:"queueGroup"`
-	NATSConfig NATSConfig    `mapstructure:"nats"`
+// Config contains event provider configs.
+type Config struct {
+	NATS NatsConfig `mapstructure:"nats"`
 }
 
-// NATSConfig handles reading in all pubsub values specific to NATS
-type NATSConfig struct {
-	Token     string `mapstructure:"token"`
-	CredsFile string `mapstructure:"credsFile"`
+// MustViperFlags returns the cobra flags and viper config for events.
+func MustViperFlags(v *viper.Viper, flags *pflag.FlagSet, appName string) {
+	MustViperFlagsForNats(v, flags, appName)
 }
 
-// MustViperFlagsForPublisher returns the cobra flags and viper config for an event publisher
-func MustViperFlagsForPublisher(v *viper.Viper, flags *pflag.FlagSet, appName string) {
-	flags.String("events-publisher-url", "nats://nats:4222", "nats server connection url")
-	viperx.MustBindFlag(v, "events.publisher.url", flags.Lookup("events-publisher-url"))
+// Option configures a connection option.
+type Option func(config *Config) error
 
-	v.MustBindEnv("events.publisher.timeout")
-	v.MustBindEnv("events.publisher.prefix")
-	v.MustBindEnv("events.publisher.source")
-	v.MustBindEnv("events.publisher.nats.token")
-	v.MustBindEnv("events.publisher.nats.credsFile")
+// WithLogger sets the logger for the connection.
+func WithLogger(logger *zap.SugaredLogger) Option {
+	return func(config *Config) error {
+		config.NATS.Logger = logger
 
-	v.SetDefault("events.publisher.timeout", defaultTimeout)
-	v.SetDefault("events.publisher.source", appName)
+		return nil
+	}
 }
 
-// MustViperFlagsForSubscriber returns the cobra flags and viper config for an event subscriber
-func MustViperFlagsForSubscriber(v *viper.Viper, flags *pflag.FlagSet) {
-	flags.String("events-subscriber-url", "nats://nats:4222", "nats server connection url")
-	viperx.MustBindFlag(v, "events.subscriber.url", flags.Lookup("events-subscriber-url"))
-	flags.String("events-subscriber-queuegroup", "", "subscriber queue group")
-	viperx.MustBindFlag(v, "events.subscriber.queueGroup", flags.Lookup("events-subscriber-queuegroup"))
+// WithNatsOptions configures nats options.
+func WithNatsOptions(options ...NatsOption) Option {
+	return func(config *Config) error {
+		var err error
 
-	v.MustBindEnv("events.subscriber.timeout")
-	v.MustBindEnv("events.subscriber.prefix")
-	v.MustBindEnv("events.subscriber.nats.token")
-	v.MustBindEnv("events.subscriber.nats.credsFile")
+		for _, opt := range options {
+			err = multierr.Append(err, opt(&config.NATS))
```

Comment:
[errors.Join](https://pkg.go.dev/errors#Join) also exists if sticking to the standard library is preferred

Reply:
ooh, good to know. I've liked the additional options multierr has given, but will see how this could replace some of it. Thanks!